webapp/security/views.py

```diff
@@ -539,3 +594,54 @@ def delete_cve(cve_id):
         return flask.jsonify({"message": error.orig.args[0]}), 400
 
     return flask.jsonify({"message": "CVE deleted succesfully"}), 200
+
+
+@authorization_required
+def bulk_upsert_cve():
+    """
+    Receives a PUT request from load_cve.py
+    Parses the object and bulk inserts or updates
+    @returns 3 lists of CVEs, created CVEs, updated CVEs and failed CVEs
+    """
+
+    cves_data = flask.request.json
+    cves_to_create = []
+    cves_to_update = []
+
+    if not type(cves_data) == list:
+        return (
+            flask.jsonify(
+                {"message": "Please, submit a list (array) of CVEs"}
+            ),
+            400,
+        )
+
+    for data in cves_data:
+        cve = db_session.query(CVE).get(data["id"])
+        if cve:
+            cves_to_update.append(data)
```

Comment:
Is adding this stuff to an array before calling `db_session.add_all` on it significantly more performant than simply calling `db_session.add()` within this for loop? It seems clear to me that the way SQLAlchemy deliberately separates the `add` step from the `commit` step is presumably specifically so you can build up your model piece by piece before you commit.

There's also significant duplication between the `update` and `create` functions.

Wouldn't this code be rather more simple if it were something like:

``` python3
for data in cves_data:
    cve = db_session.query(CVE).get(data["id"])

    if not cve:
        cve = CVE(data["id"])
    
    cve = update_cve_object(cve, data)

    db_session.add(cve)

try:
    db_session.commit()
```

Reply:
No, the `add` does not add bulk (list) of objects, it only adds one.

Reply:
> No, the add does not add bulk (list) of objects, it only adds one, it expects a single object, it will fail with this condition.

OK, so simply change my code above to say `add_all` instead of `add`.

>  It's not really about performance but about transaction, we would be doing 1 transaction per CVE if we use add().

I don't think this is true, I think the transaction is what happens when you do `db_session.commit()`. You `add` as many times as you like, I would assume including as many `add_all`s as you like, then you `commit`.

> And as mentioned before, to update we require the original object, while to create we do not need the original object, so there will be some degree of duplication (that's why I did 2 add_all).

My point here is that, if you have a new CVE, you can simply create a blank one with `CVE(data["id"])` and then update it exactly the same way you would if it already existed (see my code above). That way you don't need to duplicate any code at all.